### PR TITLE
bug: LT-1443 - Refactor Notifications

### PR DESCRIPTION
**Pull request by andytaylorkainos**

### CODE REVIEW

File: users/models.py [hunk 1]
```diff
@@ -94,23 +95,31 @@ def send_notification(self, **kwargs):
         pass
 
 
+class BaseNotification(models.Model):
+    user = models.ForeignKey(BaseUser, on_delete=models.CASCADE, null=False)
```
Comment (cstobart) [inline users/models.py]: Could this become GovUser and change organisation to the user-organisation-relationship? Unless there is a plan to put notifications on the organisation switch screen at a post map date?

Comment (andytaylorkainos) [reply users/models.py]: I had thought about doing this, but i wanted to hold the user at the parent model so that we can query all notifications based on the user - although thinking about it more, a gov user will never have an exporter user notification 🤔

File: applications/migrations/0002_baseapplication_countryonapplication_externallocationonapplication_goodonapplication_hmrcquery_opena.py [hunk 2]
```diff
@@ -10,8 +10,8 @@ class Migration(migrations.Migration):
     initial = True
 
     dependencies = [
-        ("applications", "0001_initial"),
         ("organisations", "0001_initial"),
+        ("applications", "0001_initial"),
```
Comment (BenAndrew-Kainos) [inline applications/migrations/0002_baseapplication_countryonapplication_externallocationonapplication_goodonapplication_hmrcquery_opena.py]: Why has this migration been rebuilt (no changes)

File: applications/migrations/0003_auto_20191216_1710.py [hunk 3]
```diff
@@ -9,14 +9,14 @@ class Migration(migrations.Migration):
     initial = True
 
     dependencies = [
+        ("organisations", "0001_initial"),
+        ("parties", "0001_initial"),
```
Comment (BenAndrew-Kainos) [inline applications/migrations/0003_auto_20191216_1710.py]: Why has this migration been updated?

File: applications/tests/tests_documents.py [hunk 4]
```diff
@@ -51,25 +51,24 @@ def test_upload_document_on_unsubmitted_application(self, mock_prepare_doc):
...
-        self.assertEqual(self.data["name"], document1["name"])
-        self.assertEqual(self.data["s3_key"], document1["s3_key"])
-        self.assertEqual(self.data["size"], document1["size"])
-        self.assertEqual(self.data["description"], document1["description"])
-
-        self.assertEqual(self.data2["name"], document2["name"])
-        self.assertEqual(self.data2["s3_key"], document2["s3_key"])
-        self.assertEqual(self.data2["size"], document2["size"])
-        self.assertEqual(self.data2["description"], document2["description"])
+        self.assertEqual(len(response_data), 3)
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_documents.py]: Why are you asserting that 3 documents are found (data only contains 2)

File: cases/generated_documents/migrations/0001_initial.py [hunk 5]
```diff
@@ -9,8 +9,8 @@ class Migration(migrations.Migration):
     initial = True
 
     dependencies = [
-        ("letter_templates", "0001_initial"),
         ("cases", "0001_initial"),
+        ("letter_templates", "0001_initial"),
```
Comment (BenAndrew-Kainos) [inline cases/generated_documents/migrations/0001_initial.py]: Same as above (why has this migration been changed?)

File: cases/libraries/delete_notifications.py [hunk 6]
```diff
@@ -0,0 +1,6 @@
+from users.models import ExporterNotification
+
+
+def delete_exporter_notifications(user, organisation, objects):
```
Comment (BenAndrew-Kainos) [inline cases/libraries/delete_notifications.py]: Could you specify types and rename objects to something more meaningful

File: cases/views/case_notes.py [hunk 7]
```diff
@@ -20,11 +20,13 @@ class CaseNoteList(APIView):
     def get(self, request, pk):
         """ Gets all case notes. """
         case = get_case(pk)
-        case_notes = get_case_notes_from_case(case, isinstance(request.user, ExporterUser))
-        serializer = CaseNoteSerializer(case_notes, many=True)
-
-        mark_notifications_as_viewed(request.user, case_notes)
+        if isinstance(request.user, ExporterUser):
+            case_notes = get_case_notes_from_case(case, True)
```
Comment (BenAndrew-Kainos) [inline cases/views/case_notes.py]: Could you label the arguments (I can't tell what True/False means here)

File: goods/migrations/0002_auto_20191216_1710.py [hunk 8]
```diff
@@ -9,9 +9,9 @@ class Migration(migrations.Migration):
     initial = True
 
     dependencies = [
-        ("users", "0001_initial"),
         ("goods", "0001_initial"),
         ("organisations", "0001_initial"),
+        ("users", "0001_initial"),
```
Comment (BenAndrew-Kainos) [inline goods/migrations/0002_auto_20191216_1710.py]: Same as above (why has this migration been changed?)

File: users/models.py [hunk 9]
```diff
@@ -94,23 +95,31 @@ def send_notification(self, **kwargs):
...
+
+    # All notifications are currently linked to a case
+    case = models.ForeignKey("cases.Case", on_delete=models.CASCADE, null=False)
```
Comment (BenAndrew-Kainos) [inline users/models.py]: Could you use `Case` instead of `"cases.Case"` please

File: users/views.py [hunk 10]
```diff
@@ -140,29 +142,31 @@ def get(self, request):
...
-            | Q(ecju_query__case__organisation_id=organisation_id)
-            | Q(query__organisation__id=organisation_id)
-            | Q(generated_case_document__case__organisation__id=organisation_id)
-        )
 
-        if self.request.GET.get("unviewed"):
-            queryset = queryset.filter(viewed_at__isnull=True)
+    def get(self, request):
+        data = {}
+        queryset = ExporterNotification.objects.filter(user=request.user, organisation=request.user.organisation)
```
Comment (BenAndrew-Kainos) [inline users/views.py]: Please define queryset in the class

File: applications/migrations/0002_baseapplication_countryonapplication_externallocationonapplication_goodonapplication_hmrcquery_opena.py (hunk 2 above)
Comment (andytaylorkainos) [reply applications/migrations/0002_baseapplication_countryonapplication_externallocationonapplication_goodonapplication_hmrcquery_opena.py]: Had to rebuild migrations for changing Foreign Key types (will discuss in tech share)

File: applications/migrations/0003_auto_20191216_1710.py (hunk 3 above)
Comment (andytaylorkainos) [reply applications/migrations/0003_auto_20191216_1710.py]: Had to rebuild migrations for changing Foreign Key types (will discuss in tech share)

File: cases/generated_documents/migrations/0001_initial.py (hunk 5 above)
Comment (andytaylorkainos) [reply cases/generated_documents/migrations/0001_initial.py]: Had to rebuild migrations for changing Foreign Key types (will discuss in tech share)

File: goods/migrations/0002_auto_20191216_1710.py (hunk 8 above)
Comment (andytaylorkainos) [reply goods/migrations/0002_auto_20191216_1710.py]: Had to rebuild migrations for changing Foreign Key types (will discuss in tech share)

File: applications/tests/tests_documents.py (hunk 4 above)
Comment (andytaylorkainos) [reply applications/tests/tests_documents.py]: I updated this test because it was failing as the result set was not ordered. The result set does contain 3 (the document created when creating the application in the test client helper)

File: users/models.py (hunk 9 above)
Comment (andytaylorkainos) [reply users/models.py]: I cant do this, this approach is being used to avoid circular dependencies

Comment (BenAndrew-Kainos) [reply users/models.py]: Could you instead add the import in the model then, to make clear that there is a circular dependency

Comment (andytaylorkainos) [reply users/models.py]: This is how we handle circular dependencies in our models 
Screen Shot 2019-12-17 at 13 39 55

File: cases/views/case_notes.py (hunk 7 above)
Comment (andytaylorkainos) [reply cases/views/case_notes.py]: Done!

File: cases/libraries/delete_notifications.py (hunk 6 above)
Comment (andytaylorkainos) [reply cases/libraries/delete_notifications.py]: I can't really name objects to something more meaningful because it corresponds to the model - but i will add types

File: users/views.py (hunk 10 above)
Comment (andytaylorkainos) [reply users/views.py]: Done

File: users/models.py (hunk 1 above)
Comment (andytaylorkainos) [reply users/models.py]: Decided against this, its a bit more extensible this way for allowing user-only (outside of organisation) notifications in the future

File: applications/serializers/hmrc_query.py [hunk 11]
```diff
@@ -58,6 +60,33 @@ def get_supporting_documentation(self, application):
         documents = ApplicationDocument.objects.filter(application=application)
         return ApplicationDocumentSerializer(documents, many=True).data
 
+    def get_exporter_user_notifications_count(self, instance):
```
Comment (Mark-Spiers-kainos) [inline applications/serializers/hmrc_query.py]: I just seen you had to do this function 3 times, because the class inherits from the ApplicationList serialiser. Same as some of the meta fields and functions being applied. It may be worthwhile to consider making a GenericApplicationViewSerializer, to inherit these from and cut down the code a bit (in the future)?

Comment (andytaylorkainos) [reply applications/serializers/hmrc_query.py]: Yeah, good shout! ill work on it now :)

Comment (andytaylorkainos) [reply applications/serializers/hmrc_query.py]: Done

**Comment by andytaylorkainos** [comment]
@jfaracik Good spot, changed it accordingly

File: applications/serializers/generic_application.py [hunk 12]
```diff
@@ -57,19 +80,31 @@ def get_status(self, instance):
...
-            "name",
-            "organisation",
-            "application_type",
-            "export_type",
-            "created",
-            "modified",
-            "submitted_at",
-            "status",
-            "case",
+        fields = GenericApplicationListSerializer.Meta.fields
```
Comment (Mark-Spiers-kainos) [inline applications/serializers/generic_application.py]: additional_documents and good_locations can be moved up to here.

File: audit_trail/managers.py [hunk 13]
```diff
@@ -18,11 +19,21 @@ def create(self, *args, **kwargs):
...
-        audit = super(AuditManager, self).create(*args, **kwargs)
+        target = kwargs.get("target")
 
-        if isinstance(kwargs.get("target"), Case) and isinstance(kwargs.get("actor"), ExporterUser):
-            # Notify gov users when exporter updates a case
-            for gov_user in GovUser.objects.all():
-                gov_user.send_notification(audit=audit)
+        if isinstance(target, Case):
+            # Only audit cases if they do not have status set to 'draft'
+            if target.status.status != CaseStatusEnum.DRAFT:
```
Comment (Mark-Spiers-kainos) [inline audit_trail/managers.py]: `if not is_case_status_draft(target.status):` or `if not is_case_status_draft(target.status.status): ` (can handle both) for readability

File: applications/serializers/generic_application.py (hunk 12 above)
Comment (andytaylorkainos) [reply applications/serializers/generic_application.py]: Trying not to refactor the world whilst im at this; ill do this quick change tho

File: audit_trail/managers.py (hunk 13 above)
Comment (andytaylorkainos) [reply audit_trail/managers.py]: oh, aweosme - done!

File: applications/serializers/generic_application.py [hunk 14]
```diff
@@ -57,21 +81,67 @@ def get_status(self, instance):
...
+        override this function in child classes
+        """
+
+        return get_exporter_user_notifications_total_count(
+            exporter_user=self.context.get("exporter_user"), case=instance
+        )
+
+
+class GenericApplicationViewSerializer(GenericApplicationListSerializer):
+    # TODO: Rename to supporting_documentation when possible
```
Comment (BenAndrew-Kainos) [inline applications/serializers/generic_application.py]: Could you not do this rename now (seems like a decent time with the DB cleardown and all)

File: cases/libraries/delete_notifications.py [hunk 15]
```diff
@@ -0,0 +1,7 @@
+from organisations.models import Organisation
+from users.models import ExporterNotification, ExporterUser
+
+
+def delete_exporter_notifications(user: ExporterUser, organisation: Organisation, objects: [object]):
```
Comment (BenAndrew-Kainos) [inline cases/libraries/delete_notifications.py]: This data type enforcement if pointless (everything's an object). Please just enforce this is a list. Also please could you pass a list of id's instead

File: goods/serializers.py [hunk 16]
```diff
@@ -111,10 +125,20 @@ def get_case_id(self, instance):
...
+    def get_query(self, instance):
+        query = {}
         clc_query = ControlListClassificationQuery.objects.filter(good=instance)
+
         if clc_query:
-            return clc_query.first().id
+            clc_query = clc_query.first()
+            query = {
+                "id": clc_query.id,
+                "exporter_user_notifications_count": get_exporter_user_notifications_individual_counts(
```
Comment (BenAndrew-Kainos) [inline goods/serializers.py]: Could you move this code to a method (it's used in a few places)

File: users/libraries/notifications.py [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+from users.models import ExporterNotification, ExporterUser
+from cases.models import Case
+
+
+def get_exporter_user_notifications_total_count(exporter_user: ExporterUser, case: Case) -> {}:
+    notifications_total_count = {"total": 0}
+
+    if exporter_user:
+        notifications_total_count["total"] = ExporterNotification.objects.filter(
```
Comment (BenAndrew-Kainos) [inline users/libraries/notifications.py]: Woul you mind changing this function into an if else instead returning `{"total": 0}` or `{"total": notifications_total_count}`

File: users/libraries/notifications.py [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+
+    if exporter_user:
+        queryset = (
+            ExporterNotification.objects.filter(user=exporter_user, organisation=exporter_user.organisation, case=case)
+            .values("content_type__model")
+            .annotate(count=Count("content_type__model"))
+        )
+
+        for content_type in queryset:
+            notifications_individual_counts[content_type["content_type__model"]] = content_type["count"]
```
Comment (BenAndrew-Kainos) [inline users/libraries/notifications.py]: Could you comment on this in a docstring (I'm a bit lost)

File: users/views.py (hunk 10 above)
Comment (BenAndrew-Kainos) [reply users/views.py]: This still isn't coming up as being defined in the class

File: users/views.py [hunk 19]
```diff
@@ -140,29 +142,32 @@ def get(self, request):
...
 
-        if self.request.GET.get("unviewed"):
-            queryset = queryset.filter(viewed_at__isnull=True)
+        case_types = request.GET.getlist("case_type")
+        if case_types:
+            queries = [Q(case__type=case_type) for case_type in case_types]
+            queryset = queryset.filter(reduce(or_, queries))
```
Comment (BenAndrew-Kainos) [inline users/views.py]: What does this line do (please comment)

File: users/views.py [hunk 20]
```diff
@@ -172,18 +177,13 @@ def get(self, request):
...
-                user=user,
-                audit__target_object_id=case.id,
-                audit__target_content_type=ContentType.objects.get_for_model(case),
-            )
-        except Notification.DoesNotExist:
-            return JsonResponse(data={"notification": None})
+            content_type = ContentType.objects.get_for_model(Audit)
+            notification = GovNotification.objects.get(user=user, content_type=content_type, case__id=case)
+        except GovNotification.DoesNotExist:
+            return JsonResponse(data={"notification": None}, status=status.HTTP_200_OK)
```
Comment (BenAndrew-Kainos) [inline users/views.py]: If you fetch a notification and it's not found, surely it should be a `404`

File: cases/libraries/delete_notifications.py (hunk 15 above)
Comment (andytaylorkainos) [reply cases/libraries/delete_notifications.py]: Yeah i know its pointless haha - you wanted everything typed on your previous request :)
I'll change it to a list

File: applications/serializers/generic_application.py (hunk 14 above)
Comment (andytaylorkainos) [reply applications/serializers/generic_application.py]: I'm not doing this now; i know it needs done - but the other 5 pointer in the sprint needs done more. The only reason this code is being touched in this PR was because of a change made in another PR request (one which was already out of the scope of this PR)

File: users/libraries/notifications.py (hunk 18 above)
Comment (andytaylorkainos) [reply users/libraries/notifications.py]: Done!

File: users/views.py (hunk 19 above)
Comment (andytaylorkainos) [reply users/views.py]: Done!

File: users/views.py (hunk 20 above)
Comment (andytaylorkainos) [reply users/views.py]: Yep good point, rejigged the endpoint a little

File: goods/serializers.py (hunk 16 above)
Comment (andytaylorkainos) [reply goods/serializers.py]: Done!

File: cases/libraries/get_case_note.py [hunk 21]
```diff
@@ -12,12 +12,12 @@ def get_case_note(pk):
...
 
 
-def get_case_notes_from_case(case, only_show_notes_visible_to_exporter):
+def get_case_notes_from_case(case, only_visible_to_exporter=False):
     """
     Returns all the case notes from a case
     If is_visible_to_exporter is True, then only show case notes that are visible to exporters
     """
-    if only_show_notes_visible_to_exporter:
+    if only_visible_to_exporter:
```
Comment (stephanieboyd) [inline cases/libraries/get_case_note.py]: I preferred the old name, read better in this expression

File: cases/libraries/get_case_note.py [hunk 22]
```diff
@@ -12,12 +12,12 @@ def get_case_note(pk):
         raise NotFoundError({"case_note": "Case Note not found"})
 
 
-def get_case_notes_from_case(case, only_show_notes_visible_to_exporter):
+def get_case_notes_from_case(case, only_visible_to_exporter=False):
     """
     Returns all the case notes from a case
     If is_visible_to_exporter is True, then only show case notes that are visible to exporters
```
Comment (stephanieboyd) [inline cases/libraries/get_case_note.py]: Docstring needs updating

File: goods/libraries/get_goods.py [hunk 23]
```diff
@@ -32,3 +35,18 @@ def get_good_with_organisation(pk, organisation):
...
+    if clc_query:
+        clc_query = clc_query.first()
+        return {
+            "id": clc_query.id,
+            "exporter_user_notifications_count": get_exporter_user_notifications_total_count(
+                exporter_user=exporter_user, case=clc_query
+            ),
+        }
+
+    return {}
```
Comment (stephanieboyd) [inline goods/libraries/get_goods.py]: Returning None (explicit or not, don't mind), and handling it on the other side would be nicer

File: queries/end_user_advisories/serializers.py [hunk 24]
```diff
@@ -39,6 +45,7 @@ class Meta:
             "contact_job_title",
             "contact_telephone",
             "status",
+            "exporter_user_notifications_count",
```
Comment (stephanieboyd) [inline queries/end_user_advisories/serializers.py]: I feel like this should be singluar

File: users/libraries/notifications.py [hunk 25]
```diff
@@ -0,0 +1,37 @@
+from django.db.models import Count
+
+from users.models import ExporterNotification, ExporterUser
+from cases.models import Case
+
+
+def get_exporter_user_notifications_total_count(exporter_user: ExporterUser, case: Case) -> {}:
```
Comment (stephanieboyd) [inline users/libraries/notifications.py]: Use `Dict` instead of {}

File: users/libraries/notifications.py [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+    if exporter_user:
+        return {
+            "total": ExporterNotification.objects.filter(
+                user=exporter_user, organisation=exporter_user.organisation, case=case
+            ).count()
+        }
+
+    return {"total": 0}
```
Comment (stephanieboyd) [inline users/libraries/notifications.py]: I'd define this at the top as the default return value - and then inside the condition, just update the loop. It's not the easiest to read a dictionary that's being constructed when the value is some chained operation

File: users/libraries/notifications.py [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def get_exporter_user_notifications_individual_counts(exporter_user: ExporterUser, case: Case) -> {}:
```
Comment (stephanieboyd) [inline users/libraries/notifications.py]: Same again with the return type

File: users/libraries/notifications.py [hunk 28]
```diff
@@ -0,0 +1,37 @@
...
+        # Get the total number of notifications for each type
+        queryset = (
+            ExporterNotification.objects.filter(user=exporter_user, organisation=exporter_user.organisation, case=case)
+            .values("content_type__model")
+            .annotate(count=Count("content_type__model"))
+        )
+
+        # Iterate through the different types and append to notifications_individual_counts dictionary.
+        # Set the model name as the key and the count as the value.
+        # Whilst iterating, calculate the total number of notifications instead of executing another query.
```
Comment (stephanieboyd) [inline users/libraries/notifications.py]: Unnecessary comment :)

File: users/libraries/notifications.py [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+        # Group by content_type (casenote, ecjuquery, generatedcasedocument).
+        # Get the total number of notifications for each type
+        queryset = (
+            ExporterNotification.objects.filter(user=exporter_user, organisation=exporter_user.organisation, case=case)
+            .values("content_type__model")
+            .annotate(count=Count("content_type__model"))
+        )
+
+        # Iterate through the different types and append to notifications_individual_counts dictionary.
+        # Set the model name as the key and the count as the value.
```
Comment (stephanieboyd) [inline users/libraries/notifications.py]: Maybe just an example {"model": 12, "other model": 3}

File: users/libraries/notifications.py [hunk 30]
```diff
@@ -0,0 +1,37 @@
...
+        # Whilst iterating, calculate the total number of notifications instead of executing another query.
+        for content_type in queryset:
+            notifications_individual_counts[content_type["content_type__model"]] = content_type["count"]
```
Comment (stephanieboyd) [inline users/libraries/notifications.py]: individual_model_notification_counts?

Comment (stephanieboyd) [reply users/libraries/notifications.py]: notification_counts_by_model?

File: users/models.py [hunk 31]
```diff
@@ -94,23 +95,31 @@ def send_notification(self, **kwargs):
...
+
+    # Generic Foriegn Key Fields
+    content_type = models.ForeignKey(ContentType, default=None, on_delete=models.CASCADE)
+    object_id = models.UUIDField(default=uuid.uuid4)
```
Comment (stephanieboyd) [inline users/models.py]: Not loving these 'object' names

File: users/tests/tests_exporter_user_notifications.py [hunk 32]
```diff
@@ -1,15 +1,14 @@
...
 from rest_framework import status
 
-from cases.models import Notification
+from cases.models import CaseNote
+from users.models import ExporterNotification
 from test_helpers.clients import DataTestClient
 from users.libraries.user_to_token import user_to_token
 
 
 class NotificationTests(DataTestClient):
```
Comment (stephanieboyd) [inline users/tests/tests_exporter_user_notifications.py]: Please extend name to include Exporter

File: users/tests/tests_exporter_user_notifications.py [hunk 33]
```diff
@@ -18,7 +17,7 @@ def tests_create_new_clc_query_notification(self):
         self.create_case_note(clc_case, "This is a test note 3", self.gov_user, True)
         self.create_case_note(clc_case, "This is a test note 4", self.gov_user, False)
```
Comment (stephanieboyd) [inline users/tests/tests_exporter_user_notifications.py]: Are these case notes relevant to the test?

File: users/tests/tests_exporter_user_notifications.py [hunk 34]
```diff
@@ -28,29 +27,45 @@ def test_create_new_application_notification(self):
...
+        clc_case = self.create_clc_query("Example CLC Query", self.organisation)
         self.create_case_note(clc_case, "This is a test note 1", self.gov_user, True)
         self.create_case_note(clc_case, "This is a test note 2", self.gov_user, True)
         self.create_case_note(clc_case, "This is a test note 3", self.gov_user, True)
+        case_note_content_type = ContentType.objects.get_for_model(CaseNote)
 
-        self.assertEqual(Notification.objects.all().count(), 7)
         self.assertEqual(
-            Notification.objects.filter(case_note__case__id=case.id).count(), 4,
+            ExporterNotification.objects.filter(
```
Comment (stephanieboyd) [inline users/tests/tests_exporter_user_notifications.py]: Please extract out into named variables and then the asserts will be easy to read.

File: users/tests/tests_gov_user_case_notifications.py [hunk 35]
```diff
@@ -4,39 +4,34 @@
 
 from audit_trail.models import Audit
 from audit_trail.payload import AuditType
-from cases.models import Notification
+from users.models import GovNotification
 from test_helpers.clients import DataTestClient
 
 
 class NotificationTests(DataTestClient):
```
Comment (stephanieboyd) [inline users/tests/tests_gov_user_case_notifications.py]: Please extend name with Gov

File: users/tests/tests_gov_user_case_notifications.py [hunk 36]
```diff
@@ -45,34 +40,28 @@ def tests_edit_application_updates_previous_case_notification_success(self):
...
+            user=self.gov_user, content_type=content_type, case=case
         ).count()
 
-        url = reverse("applications:application", kwargs={"pk": application.id})
+        url = reverse("applications:application", kwargs={"pk": case.id})
         data = {"name": "even newer app name!"}
 
         response = self.client.put(url, data, **self.exporter_headers)
-        application.refresh_from_db()
+        case.refresh_from_db()
```
Comment (stephanieboyd) [inline users/tests/tests_gov_user_case_notifications.py]: Why twice?

File: users/tests/tests_gov_user_case_notifications.py [hunk 37]
```diff
@@ -81,39 +70,34 @@ def tests_get_case_notification_deletes_case_notification_and_returns_data(self)
...
         response = self.client.get(url, **self.gov_headers)
         notification = response.json()["notification"]
 
         self.assertEqual(len(notification), 1)
-        self.assertEqual(int(notification["audit_id"]), audit.id)
+        self.assertEqual(notification["audit_id"], str(audit.id))
 
-        self.assertEqual(Notification.objects.filter(user=self.gov_user, audit=audit).count(), 0)
+        self.assertEqual(
+            GovNotification.objects.filter(user=self.gov_user, content_type=content_type, case=case).count(), 0,
```
Comment (stephanieboyd) [inline users/tests/tests_gov_user_case_notifications.py]: Please pull out into variables, asserts should be easy to read

Comment (stephanieboyd) [reply users/tests/tests_gov_user_case_notifications.py]: In other places too :)

File: users/tests/tests_gov_user_case_notifications.py [hunk 38]
```diff
@@ -81,39 +70,34 @@ def tests_get_case_notification_deletes_case_notification_and_returns_data(self)
...
+        )
 
     def tests_edit_application_as_gov_user_does_not_create_a_case_notification(self):
```
Comment (stephanieboyd) [inline users/tests/tests_gov_user_case_notifications.py]: I dunno if we're bothering but success/failure?

File: users/views.py [hunk 39]
```diff
@@ -140,50 +142,52 @@ def get(self, request):
...
+        # to get notifications matching different case types
+        case_types = request.GET.getlist("case_type")
+        if case_types:
+            queries = [Q(case__type=case_type) for case_type in case_types]
+            # Collapses the queries list into a usable filter
+            queryset = queryset.filter(reduce(or_, queries))
 
-        if self.request.GET.get("unviewed"):
-            queryset = queryset.filter(viewed_at__isnull=True)
+        # Count the number of notifications for each type
```
Comment (stephanieboyd) [inline users/views.py]: Don't we have a function for this, in this PR?

File: users/libraries/notifications.py (hunk 17 above)
Comment (andytaylorkainos) [reply users/libraries/notifications.py]: I think ive addressed this

File: applications/serializers/generic_application.py [hunk 40]
```diff
@@ -57,21 +81,66 @@ def get_status(self, instance):
     def get_case(self, instance):
         return instance.pk
 
+    @abc.abstractmethod
+    def get_exporter_user_notifications_count(self, instance):
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply applications/serializers/generic_application.py]: it contains a list of notifications with their count tho

File: audit_trail/managers.py [hunk 41]
```diff
@@ -18,11 +19,21 @@ def create(self, *args, **kwargs):
...
 
-        audit = super(AuditManager, self).create(*args, **kwargs)
+        target = kwargs.get("target")
 
-        if isinstance(kwargs.get("target"), Case) and isinstance(kwargs.get("actor"), ExporterUser):
-            # Notify gov users when exporter updates a case
-            for gov_user in GovUser.objects.all():
-                gov_user.send_notification(audit=audit)
+        if isinstance(target, Case):
+            # Only audit cases if they do not have status set to 'draft'
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply audit_trail/managers.py]: Done

File: applications/tests/tests_standard_application_submit.py [hunk 42]
```diff
@@ -27,6 +28,8 @@ def test_submit_standard_application_success(self):
         self.assertEqual(case.status.status, CaseStatusEnum.SUBMITTED)
         for good_on_application in GoodOnApplication.objects.filter(application=case):
             self.assertEqual(good_on_application.good.status, GoodStatus.SUBMITTED)
+        # Unsubmitted (draft) applications should not create audit entries when submitted for the first time
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply applications/tests/tests_standard_application_submit.py]: Done

File: applications/views/applications.py [hunk 43]
```diff
@@ -98,7 +101,7 @@ def get(self, request, application):
         Retrieve an application instance
         """
         serializer = get_application_view_serializer(application)
-        serializer = serializer(application)
+        serializer = serializer(application, context={"exporter_user": request.user})
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply applications/views/applications.py]: get_serializer_context is not in that class

File: applications/views/applications.py [hunk 44]
```diff
@@ -193,7 +196,7 @@ def put(self, request, application):
 
         data = {"application": {**serializer.data}}
 
-        if previous_application_status:
+        if not is_case_status_draft(previous_application_status.status):
             # If the application is being submitted after being edited
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply applications/views/applications.py]: updated

File: cases/libraries/get_case_note.py (hunk 21 above)
Comment (andytaylorkainos) [reply cases/libraries/get_case_note.py]: reverted

File: cases/libraries/get_case_note.py (hunk 22 above)
Comment (andytaylorkainos) [reply cases/libraries/get_case_note.py]: updated

File: goods/libraries/get_goods.py (hunk 23 above)
Comment (andytaylorkainos) [reply goods/libraries/get_goods.py]: Think ive addressed this

File: users/libraries/notifications.py (hunk 25 above)
Comment (andytaylorkainos) [reply users/libraries/notifications.py]: done

File: users/libraries/notifications.py (hunk 26 above)
Comment (andytaylorkainos) [reply users/libraries/notifications.py]: done

File: users/libraries/notifications.py (hunk 27 above)
Comment (andytaylorkainos) [reply users/libraries/notifications.py]: done

File: users/libraries/notifications.py (hunk 28 above)
Comment (andytaylorkainos) [reply users/libraries/notifications.py]: This is outdated

File: users/libraries/notifications.py (hunk 29 above)
Comment (andytaylorkainos) [reply users/libraries/notifications.py]: Done

File: users/models.py (hunk 31 above)
Comment (andytaylorkainos) [reply users/models.py]: i tried using our own names and it didnt work, it seems to use generic foreign keys theres an issue with the naming convention so i stuck with the documentation

File: users/tests/tests_exporter_user_notifications.py (hunk 32 above)
Comment (andytaylorkainos) [reply users/tests/tests_exporter_user_notifications.py]: done

File: users/tests/tests_gov_user_case_notifications.py (hunk 35 above)
Comment (andytaylorkainos) [reply users/tests/tests_gov_user_case_notifications.py]: done

File: users/tests/tests_gov_user_case_notifications.py (hunk 36 above)
Comment (andytaylorkainos) [reply users/tests/tests_gov_user_case_notifications.py]: no idea, removed

File: users/views.py (hunk 39 above)
Comment (andytaylorkainos) [reply users/views.py]: this is grouping by `case_type` where-as the function groups by `content_type` (casenote, ecjuquery etc.) 
This endpoint is currently only being used on the exporter hub page

File: users/tests/tests_exporter_user_notifications.py (hunk 33 above)
Comment (andytaylorkainos) [reply users/tests/tests_exporter_user_notifications.py]: The "false" param creates case notes that are not visible to exporter users and therefore should not raise a notification

File: users/tests/tests_gov_user_case_notifications.py (hunk 38 above)
Comment (andytaylorkainos) [reply users/tests/tests_gov_user_case_notifications.py]: done

File: queries/end_user_advisories/serializers.py (hunk 24 above)
Comment (andytaylorkainos) [reply queries/end_user_advisories/serializers.py]: discussed with other people; will rename to singular :)

File: cases/libraries/delete_notifications.py [hunk 45]
```diff
@@ -0,0 +1,7 @@
+from organisations.models import Organisation
+from users.models import ExporterNotification, ExporterUser
+
+
+def delete_exporter_notifications(user: ExporterUser, organisation: Organisation, objects: []):
```
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply cases/libraries/delete_notifications.py]: resolved on slack (its not a default, its a type)

File: users/libraries/notifications.py (hunk 30 above)
Comment (andytaylorkainos) [reply users/libraries/notifications.py]: renamed to match field (removed plural from field name)

File: applications/serializers/generic_application.py (hunk 40 above)
[the remark this answers is not in the document]
Comment (andytaylorkainos) [reply applications/serializers/generic_application.py]: renamed

File: users/tests/tests_exporter_user_notifications.py (hunk 34 above)
Comment (andytaylorkainos) [reply users/tests/tests_exporter_user_notifications.py]: done

File: applications/views/applications.py [hunk 46]
```diff
@@ -193,8 +196,8 @@ def put(self, request, application):
 
         data = {"application": {**serializer.data}}
 
-        if previous_application_status:
-            # If the application is being submitted after being edited
+        if not is_case_status_draft(previous_application_status.status):
+            # Only create the audit if the application was not previously in `Draft` status
```
Comment (stephanieboyd) [inline applications/views/applications.py]: Oh, word previously is a bit misleading here

Comment (andytaylorkainos) [reply applications/views/applications.py]: changed to `# Only create the audit if the previous application status was not 'Draft'`

File: users/tests/tests_gov_user_case_notifications.py (hunk 37 above)
Comment (andytaylorkainos) [reply users/tests/tests_gov_user_case_notifications.py]: done :)
